### DISCUSSION TITLE
Node-ffi Library blocks the launching of the electron

### DISCUSSION THREAD

**Discussion by YOSSTM** [eclipse-theia/theia#11135] [General] [answered]
I load a Node-ffi Library in Theia and then rebuild the electron and launch it , I got this error:

index.js:112 Error: No native build was found for platform=win32 arch=x64 runtime=electron abi=98 uv=1 libc=glibc node=16.5.0 electron=15.3.7 webpack=true

**Comment by vince-fugnitto** [comment]
@YOSSTM thank you for the discussion, do you have any additional info that may help? 
How and where did you load a node-ffi library?

**Reply by YOSSTM** [reply]
import ffi = require('ffi-napi');
const API= ffi.Library("path to DLL", {});

**Reply by YOSSTM** [reply]
I have tried to add a script in package.json rebuild: electron-rebuild -f -w ffi-napi && electron-rebuild -f -w ref-napi 
but the problem persists

**Comment by YOSSTM** [comment]
@vince-fugnitto Is there any solution to customize the webpack already generated by theia ?
I think I have to update the configuration of webpack to make it work(externals:{ffi:'ffi-napi,ref:ref-napi}

**Reply by vince-fugnitto** [reply]
@YOSSTM I believe you should be able to edit the `webpack.config.js` to provide any customizations:

https://github.com/eclipse-theia/theia/blob/f249b3e93787905d960d7e428f2da0c1796b76d3/examples/electron/webpack.config.js#L1-L17

The same file should exist in your own application after a build.

**Reply by YOSSTM** [reply]
Unfortunately I cannot , webpack.config.js seems as well auto-generated , tried to modify it and then rebuild the elecron yarn:build and it restores his original  content

**Reply by vince-fugnitto** [reply]
I'm suprised, it seems to work fine for me using the example in the repo. 
After a build I can modify the file and even after rebuilding the changes take effect. This is consistent what I noticed in the [changelog documentation](https://github.com/eclipse-theia/theia/blob/master/doc/changelogs/CHANGELOG-2020.md):

> [cli] renamed generated webpack config to gen-webpack.config.js https://github.com/eclipse-theia/theia/pull/6852. webpack.config.js is generated only once. It can be edited by users to customize bundling, but should be based on gen-webpack.config.js to pick any changes in the generated config. If it does not have a reference to gen-webpack.config.js then it will be regenerated.

**Reply by YOSSTM** [reply]
Yes I see, I deleted gen-webpack.config.js import from webpack config file

**Comment by YOSSTM** [comment, marked as answer]
Answer for all 👍 
Webpack block the build of native modules 
solution is to define your native modules are externals in the webpack config file
here the answer:

config.externals = {
    'ffi-napi': "require('ffi-napi')",
    'ref-napi': "require('ref-napi')",
};
